### PR TITLE
feat: Add functionalities to read and export the checklist in CSV format

### LINKED ISSUE #66: Add functionality to read and export the checklist in CSV format

### PR DESCRIPTION
**Pull request by SoloSynth1**
This PR:
- Modifies the existing YAML checklist to add an attribute `ID` for every test area and test item
- Adds two new methods `to_yaml` and `to_csv` to class `Checklist` in order to export (dump) YAML and/or CSV files
- Extends `Checklist`'s capability to read checklist items from CSV files

The CSV files must be in specific format exported by `Checklist`. Please refer to the example calls to see how to read CSV files.

Closes #66 .

### CODE REVIEW

**Comment by SoloSynth1** [comment]
@JohnShiuMK Thanks for the review. I have changed the CSV dumping behaviour after our yesterday meeting so that now it will output a `Topic` column in `tests.csv`.

As I will do some code refactor today, I plan to put this `checklist` along with `code_analyzer` into `src/test_creation/modules` so that they can be imported by other scripts without the need to set up symlinks. As such, would you mind to give it a quick review and, if it's OK, merge it into `main`? Thanks!

**Review by jinyz8888 — COMMENTED** [review]
what is format for ID?

**Comment by SoloSynth1** [comment]
Here is the current format of the output in `tests.csv` (just top 3 rows since it is too lengthy to show the entire checklist here):

| ID  | Topic            | Title                                                  | Requirement                                                                                                                                                                                  | Explanation                                                                                                                                                                                                                                 | References                |
| --- | ---------------- | ------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------- |
| 1.1 | General          | Write Descriptive Test Names                           | Each test function should have a clear, descriptive name that accurately reflects the test's purpose and the specific functionality or scenario it examines.                                 | If out tests are narrow and sufficiently descriptive, the test name itself may give us enough information to start debugging. This also helps us to identify what is being tested inside the function.                                      | trenk2014, winters2024    |
| 1.2 | General          | Keep Tests Focused                                     | Each test should focus on a single scenario, using only one set of mock data and testing one specific behavior or outcome to ensure clarity and isolate issues.                              | If we test multiple scenarios in a single test, it is hard to idenitfy exactly what went wrong. Keeping one scenario in a single test helps us to isolate problematic scenarios.                                                            | yu2018                    |
| 1.3 | General          | Prefer Narrow Assertions in Unit Tests                 | Assertions within tests should be focused and narrow. Ensure you are only testing relevant behaviors of complex objects and not including unrelated assertions.                              | If we have overly wide assertions (such as depending on every field of a complex output proto), the test may fail for many unimportant reasons. False positives are the opoosite of actionable.                                             | kent2024                  |

**Comment by SoloSynth1** [comment]
> what is format for ID?

The ID is used as a way to anchor and track the response from LLMs and will be injected into the prompt.

File: src/checklist/checklist.py [hunk 1]
```diff
@@ -39,13 +169,37 @@ def get_all_tests(self, requirements_only: bool = False):
...
+                "formatting, use `no_preserve_format=True`."
+            )
+        elif not exist_ok and os.path.exists(output_path):
+            raise FileExistsError("Output file already exists. Use `exist_ok=True` to overwrite.")
+        YamlChecklistIO.write(output_path, self.content)
+
+    def to_csv(self, output_path, exist_ok: bool = False):
+        """Dump the checklist to three separate CSV files."""
+        if not exist_ok and os.path.exists(output_path):
+            raise FileExistsError("Output file already exists. Use `exist_ok=True` to overwrite.")
```
Comment (tonyshumlh) [inline src/checklist/checklist.py:185]: Maybe it is more clear to write "Output directory already exists. Use `exist_ok=True` to overwrite."

**Comment by tonyshumlh** [comment]
Tested all functions within class `Checklist` and tested yaml->yaml,csv and csv->yaml,csv. All looks good to me. A minor comment is given

**Comment by SoloSynth1** [comment]
> Tested all functions within class `Checklist` and tested yaml->yaml,csv and csv->yaml,csv. All looks good to me. A minor comment is given

Thanks @tonyshumlh, but it seems you haven't published it yet? I am not seeing the comment.

File: src/checklist/checklist.py (hunk 1 above)
Comment (SoloSynth1) [reply src/checklist/checklist.py]: Done! Changed in latest commit 98e9373.

**Review by tonyshumlh — APPROVED** [review]

**Pull request merged**